File: javascript/src/logging.ts
/**
 * Structured logging with correlation IDs for observability.
 *
 * Provides JSON-formatted logs with pluggable output handlers.
 */

/** Log levels for structured logging. */
export enum LogLevel {
    DEBUG = "debug",
    INFO = "info",
    WARN = "warn",
    ERROR = "error",
}

/** Standard log events for IPC operations. */
export enum LogEvent {
    // Lifecycle
    WORKER_START = "worker_start",
    WORKER_STOP = "worker_stop",
    WORKER_RESTART = "worker_restart",

    // Requests
    REQUEST_START = "request_start",
    REQUEST_END = "request_end",
    REQUEST_TIMEOUT = "request_timeout",
    REQUEST_ERROR = "request_error",

    // Circuit breaker
    CIRCUIT_OPEN = "circuit_open",
    CIRCUIT_CLOSE = "circuit_close",
    CIRCUIT_HALF_OPEN = "circuit_half_open",

    // Connection
    SOCKET_CONNECT = "socket_connect",
    SOCKET_DISCONNECT = "socket_disconnect",
    SOCKET_RECONNECT = "socket_reconnect",

    // Process
    PROCESS_SPAWN = "process_spawn",
    PROCESS_EXIT = "process_exit",

    // Heartbeat
    HEARTBEAT_SENT = "heartbeat_sent",
    HEARTBEAT_RECEIVED = "heartbeat_received",
    HEARTBEAT_MISSED = "heartbeat_missed",
    HEARTBEAT_TIMEOUT = "heartbeat_timeout",

    // Queue
    QUEUE_OVERFLOW = "queue_overflow",
}

/** Structured log entry with all context. */
export interface LogEntry {
    // Required
    event: string;
    level: string;
    message: string;
    timestamp: number;

    // Correlation
    correlationId?: string;
    requestId?: string;
    parentRequestId?: string;

    // Context
    workerId?: string;
    serviceId?: string;
    func?: string;
    namespace?: string;

    // Timing
    durationMs?: number;

    // Status
    success?: boolean;
    error?: string;
    errorType?: string;

    // Metrics snapshot (optional)
    metrics?: Record<string, unknown>;

    // Custom metadata
    metadata?: Record<string, unknown>;
}

/** Type alias for log handler function. */
export type LogHandler = (entry: LogEntry) => void;

/**
 * Structured logger with pluggable handlers.
 *
 * Usage:
 *     // Simple usage with callback
 *     const logger = new StructuredLogger({
 *         handler: (entry) => console.log(JSON.stringify(entry))
 *     });
 *
 *     // Log events
 *     logger.log({
 *         event: LogEvent.REQUEST_START,
 *         message: "Calling remote function",
 *         requestId: "abc123",
 *         func: "add",
 *     });
 *
 *     // Convenience methods
 *     logger.info(LogEvent.WORKER_START, "Worker started", { workerId: "worker-1" });
 *     logger.error(LogEvent.REQUEST_ERROR, "Call failed", { error: "Timeout" });
 */
export class StructuredLogger {
    private handler?: LogHandler;
    private level: LogLevel;
    private workerId?: string;
    private serviceId?: string;

    private readonly levelOrder: Record<LogLevel, number> = {
        [LogLevel.DEBUG]: 0,
        [LogLevel.INFO]: 1,
        [LogLevel.WARN]: 2,
        [LogLevel.ERROR]: 3,
    };

    constructor(options?: {
        handler?: LogHandler;
        level?: LogLevel;
        workerId?: string;
        serviceId?: string;
    }) {
        this.handler = options?.handler;
        this.level = options?.level !== undefined ? options.level : LogLevel.INFO;
        this.workerId = options?.workerId;
        this.serviceId = options?.serviceId;
    }

    /** Set or update the log handler. */
    setHandler(handler: LogHandler): void {
        this.handler = handler;
    }

    /** Set default context for all log entries. */
    setContext(options?: { workerId?: string; serviceId?: string }): void {
        if (options?.workerId !== undefined) {
            this.workerId = options.workerId;
        }
        if (options?.serviceId !== undefined) {
            this.serviceId = options.serviceId;
        }
    }

    /** Check if this level should be logged. */
    private shouldLog(level: LogLevel): boolean {
        return this.levelOrder[level] >= this.levelOrder[this.level];
    }

    /**
     * Log an event with structured data.
     */
    log(event: LogEvent, message: string, options?: {
        level?: LogLevel;
        correlationId?: string;
        requestId?: string;
        func?: string;
        namespace?: string;
        durationMs?: number;
        success?: boolean;
        error?: string;
        errorType?: string;
        metadata?: Record<string, unknown>;
    }): void {
        if (!this.handler || !this.shouldLog(options?.level ?? LogLevel.INFO)) {
            return;
        }

        const entry: LogEntry = {
            event: event,
            level: (options?.level ?? LogLevel.INFO).valueOf(),
            message,
            timestamp: Date.now(), // Use milliseconds for consistency
            workerId: this.workerId,
            serviceId: this.serviceId,
            correlationId: options?.correlationId,
            requestId: options?.requestId,
            func: options?.func,
            namespace: options?.namespace,
            durationMs: options?.durationMs,
            success: options?.success,
            error: options?.error,
            errorType: options?.errorType,
            metadata: options?.metadata,
        };

        try {
            this.handler(entry);
        } catch (e) {
            // Don't let logging errors break the application
            console.error(`Log handler error: ${e}`);
        }
    }

    /** Log at DEBUG level. */
    debug(event: LogEvent, message: string, options?: Omit<Parameters<typeof this.log>[2], 'level'>): void {
        this.log(event, message, { ...options, level: LogLevel.DEBUG });
    }

    /** Log at INFO level. */
    info(event: LogEvent, message: string, options?: Omit<Parameters<typeof this.log>[2], 'level'>): void {
        this.log(event, message, { ...options, level: LogLevel.INFO });
    }

    /** Log at WARN level. */
    warn(event: LogEvent, message: string, options?: Omit<Parameters<typeof this.log>[2], 'level'>): void {
        this.log(event, message, { ...options, level: LogLevel.WARN });
    }

    /** Log at ERROR level. */
    error(event: LogEvent, message: string, options?: Omit<Parameters<typeof this.log>[2], 'level'>): void {
        this.log(event, message, { ...options, level: LogLevel.ERROR });
    }

    // Convenience methods for common events

    /** Log request start. */
    requestStart(options: {
        requestId: string;
        func: string;
        namespace?: string;
        correlationId?: string;
        metadata?: Record<string, unknown>;
    }): void {
        this.debug(LogEvent.REQUEST_START, `Calling ${options.func}`, {
            requestId: options.requestId,
            func: options.func,
            namespace: options.namespace ?? "default",
            correlationId: options.correlationId,
            metadata: options.metadata,
        });
    }

    /** Log request completion. */
    requestEnd(options: {
        requestId: string;
        func: string;
        durationMs: number;
        success?: boolean;
        error?: string;
        correlationId?: string;
    }): void {
        const event = options.success !== false ? LogEvent.REQUEST_END : LogEvent.REQUEST_ERROR;
        const level = options.success !== false ? LogLevel.INFO : LogLevel.WARN;
        this.log(event, options.success !== false ? `Completed ${options.func}` : `Failed ${options.func}`, {
            level,
            requestId: options.requestId,
            func: options.func,
            durationMs: Math.round(options.durationMs * 100) / 100,
            success: options.success ?? true,
            error: options.error,
            correlationId: options.correlationId,
        });
    }

    /** Log circuit breaker opening. */
    circuitOpen(failures: number): void {
        this.warn(LogEvent.CIRCUIT_OPEN, `Circuit breaker opened after ${failures} failures`);
    }

    /** Log circuit breaker closing (recovery). */
    circuitClose(): void {
        this.info(LogEvent.CIRCUIT_CLOSE, "Circuit breaker closed (recovered)");
    }

    /** Log worker start. */
    workerStart(mode: string = "spawn"): void {
        this.info(LogEvent.WORKER_START, `Worker started in ${mode} mode`);
    }

    /** Log worker stop. */
    workerStop(reason: string = "shutdown"): void {
        this.info(LogEvent.WORKER_STOP, `Worker stopped: ${reason}`);
    }

    /** Log child process exit. */
    processExit(exitCode: number): void {
        const level = exitCode === 0 ? LogLevel.INFO : LogLevel.WARN;
        this.log(LogEvent.PROCESS_EXIT, `Child process exited with code ${exitCode}`, {
            level,
            metadata: { exitCode },
        });
    }

    /** Log heartbeat sent. */
    heartbeatSent(): void {
        this.debug(LogEvent.HEARTBEAT_SENT, "Heartbeat sent");
    }

    /** Log heartbeat response received. */
    heartbeatReceived(rttMs: number): void {
        this.debug(LogEvent.HEARTBEAT_RECEIVED, `Heartbeat received (RTT: ${rttMs.toFixed(1)}ms)`, {
            durationMs: rttMs,
        });
    }

    /** Log missed heartbeat. */
    heartbeatMissed(consecutive: number, maxAllowed: number): void {
        this.warn(LogEvent.HEARTBEAT_MISSED, `Heartbeat missed (${consecutive}/${maxAllowed})`, {
            metadata: { consecutiveMisses: consecutive, maxAllowed },
        });
    }

    /** Log heartbeat timeout (too many misses). */
    heartbeatTimeout(misses: number): void {
        this.error(LogEvent.HEARTBEAT_TIMEOUT, `Heartbeat timeout after ${misses} consecutive misses`, {
            metadata: { totalMisses: misses },
        });
    }
}

/** Default handler that prints JSON to stdout. */
export function defaultJsonHandler(entry: LogEntry): void {
    console.log(JSON.stringify(entry));
}

/** Default handler that prints human-readable output. */
export function defaultPrettyHandler(entry: LogEntry): void {
    const timestamp = new Date(entry.timestamp).toLocaleTimeString("en-US", { hour12: false });
    const level = entry.level.toUpperCase().padEnd(5);
    const prefix = `[${timestamp}] [${level}]`;

    const parts: string[] = [prefix, entry.event, entry.message];

    if (entry.requestId) {
        parts.push(`req=${entry.requestId.slice(0, 8)}`);
    }
    if (entry.func) {
        parts.push(`fn=${entry.func}`);
    }
    if (entry.durationMs !== undefined) {
        parts.push(`${entry.durationMs.toFixed(1)}ms`);
    }
    if (entry.error) {
        parts.push(`error=${entry.error}`);
    }

    console.log(parts.join(" "));
}

File: rust/src/logging.rs
//! Structured logging with correlation IDs for observability.
//!
//! Provides JSON-formatted logs with pluggable output handlers.

use serde::{Deserialize, Serialize};
use std::fmt;
use std::sync::Arc;

/// Log levels for structured logging.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum LogLevel {
    Debug,
    Info,
    Warn,
    Error,
}

impl fmt::Display for LogLevel {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            LogLevel::Debug => write!(f, "debug"),
            LogLevel::Info => write!(f, "info"),
            LogLevel::Warn => write!(f, "warn"),
            LogLevel::Error => write!(f, "error"),
        }
    }
}

/// Standard log events for IPC operations.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LogEvent {
    // Lifecycle
    WorkerStart,
    WorkerStop,
    WorkerRestart,

    // Requests
    RequestStart,
    RequestEnd,
    RequestTimeout,
    RequestError,

    // Circuit breaker
    CircuitOpen,
    CircuitClose,
    CircuitHalfOpen,

    // Connection
    SocketConnect,
    SocketDisconnect,
    SocketReconnect,

    // Process
    ProcessSpawn,
    ProcessExit,

    // Heartbeat
    HeartbeatSent,
    HeartbeatReceived,
    HeartbeatMissed,
    HeartbeatTimeout,

    // Queue
    QueueOverflow,
}

impl fmt::Display for LogEvent {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = serde_json::to_string(self).unwrap_or_default();
        write!(f, "{}", s.trim_matches('"'))
    }
}

/// Structured log entry with all context.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LogEntry {
    // Required
    pub event: String,
    pub level: String,
    pub message: String,
    pub timestamp: f64,

    // Correlation
    #[serde(skip_serializing_if = "Option::is_none")]
    pub correlation_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub request_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_request_id: Option<String>,

    // Context
    #[serde(skip_serializing_if = "Option::is_none")]
    pub worker_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub service_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub function: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub namespace: Option<String>,

    // Timing
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration_ms: Option<f64>,

    // Status
    #[serde(skip_serializing_if = "Option::is_none")]
    pub success: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_type: Option<String>,

    // Metrics snapshot (optional)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metrics: Option<serde_json::Value>,

    // Custom metadata
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<serde_json::Value>,
}

impl LogEntry {
    /// Convert to JSON string.
    pub fn to_json(&self) -> String {
        serde_json::to_string(self).unwrap_or_default()
    }
}

impl Default for LogEntry {
    fn default() -> Self {
        Self {
            event: String::new(),
            level: String::new(),
            message: String::new(),
            timestamp: 0.0,
            correlation_id: None,
            request_id: None,
            parent_request_id: None,
            worker_id: None,
            service_id: None,
            function: None,
            namespace: None,
            duration_ms: None,
            success: None,
            error: None,
            error_type: None,
            metrics: None,
            metadata: None,
        }
    }
}

/// Type alias for log handler function.
pub type LogHandler = Arc<dyn Fn(&LogEntry) + Send + Sync>;

/// Structured logger with pluggable handlers.
///
/// # Example
///
/// ```rust,no_run
/// use multifrost::logging::{StructuredLogger, LogEvent, LogLevel};
/// use std::sync::Arc;
///
/// let logger = StructuredLogger::new(
///     Some(Arc::new(|entry| {
///         println!("{}", entry.to_json());
///     })),
///     LogLevel::Info,
///     Some("worker-1".to_string()),
///     None,
/// );
///
/// logger.info(LogEvent::WorkerStart, "Worker started");
/// ```
#[derive(Clone)]
pub struct StructuredLogger {
    handler: Option<LogHandler>,
    level: LogLevel,
    worker_id: Option<String>,
    service_id: Option<String>,
}

impl StructuredLogger {
    /// Create a new structured logger.
    pub fn new(
        handler: Option<LogHandler>,
        level: LogLevel,
        worker_id: Option<String>,
        service_id: Option<String>,
    ) -> Self {
        Self {
            handler,
            level,
            worker_id,
            service_id,
        }
    }

    /// Set or update the log handler.
    pub fn set_handler(&mut self, handler: LogHandler) {
        self.handler = Some(handler);
    }

    /// Set default context for all log entries.
    pub fn set_context(&mut self, worker_id: Option<String>, service_id: Option<String>) {
        if worker_id.is_some() {
            self.worker_id = worker_id;
        }
        if service_id.is_some() {
            self.service_id = service_id;
        }
    }

    /// Check if this level should be logged.
    fn should_log(&self, level: LogLevel) -> bool {
        level as u8 >= self.level as u8
    }

    /// Log an event with structured data.
    pub fn log(&self, event: LogEvent, message: &str, level: LogLevel, options: LogOptions) {
        if self.handler.is_none() || !self.should_log(level) {
            return;
        }

        let entry = LogEntry {
            event: event.to_string(),
            level: level.to_string(),
            message: message.to_string(),
            timestamp: current_timestamp(),
            worker_id: self.worker_id.clone(),
            service_id: self.service_id.clone(),
            correlation_id: options.correlation_id,
            request_id: options.request_id,
            parent_request_id: options.parent_request_id,
            function: options.function,
            namespace: options.namespace,
            duration_ms: options.duration_ms,
            success: options.success,
            error: options.error,
            error_type: options.error_type,
            metrics: options.metrics,
            metadata: options.metadata,
        };

        if let Some(ref handler) = self.handler {
            handler(&entry);
        }
    }

    /// Log at DEBUG level.
    pub fn debug(&self, event: LogEvent, message: &str, options: LogOptions) {
        self.log(event, message, LogLevel::Debug, options);
    }

    /// Log at INFO level.
    pub fn info(&self, event: LogEvent, message: &str, options: LogOptions) {
        self.log(event, message, LogLevel::Info, options);
    }

    /// Log at WARN level.
    pub fn warn(&self, event: LogEvent, message: &str, options: LogOptions) {
        self.log(event, message, LogLevel::Warn, options);
    }

    /// Log at ERROR level.
    pub fn error(&self, event: LogEvent, message: &str, options: LogOptions) {
        self.log(event, message, LogLevel::Error, options);
    }

    // Convenience methods for common events

    /// Log request start.
    pub fn request_start(
        &self,
        request_id: &str,
        function: &str,
        namespace: &str,
        correlation_id: Option<String>,
        metadata: Option<serde_json::Value>,
    ) {
        self.debug(
            LogEvent::RequestStart,
            &format!("Calling {}", function),
            LogOptions {
                request_id: Some(request_id.to_string()),
                function: Some(function.to_string()),
                namespace: Some(namespace.to_string()),
                correlation_id,
                metadata,
                ..Default::default()
            },
        );
    }

    /// Log request completion.
    pub fn request_end(
        &self,
        request_id: &str,
        function: &str,
        duration_ms: f64,
        success: bool,
        error: Option<String>,
        correlation_id: Option<String>,
    ) {
        let event = if success {
            LogEvent::RequestEnd
        } else {
            LogEvent::RequestError
        };
        let level = if success {
            LogLevel::Info
        } else {
            LogLevel::Warn
        };

        self.log(
            event,
            &format!(
                "{} {}",
                if success { "Completed" } else { "Failed" },
                function
            ),
            level,
            LogOptions {
                request_id: Some(request_id.to_string()),
                function: Some(function.to_string()),
                duration_ms: Some(duration_ms),
                success: Some(success),
                error,
                correlation_id,
                ..Default::default()
            },
        );
    }

    /// Log circuit breaker opening.
    pub fn circuit_open(&self, failures: usize) {
        self.warn(
            LogEvent::CircuitOpen,
            &format!("Circuit breaker opened after {} failures", failures),
            LogOptions::default(),
        );
    }

    /// Log circuit breaker closing (recovery).
    pub fn circuit_close(&self) {
        self.info(
            LogEvent::CircuitClose,
            "Circuit breaker closed (recovered)",
            LogOptions::default(),
        );
    }

    /// Log worker start.
    pub fn worker_start(&self, mode: &str) {
        self.info(
            LogEvent::WorkerStart,
            &format!("Worker started in {} mode", mode),
            LogOptions::default(),
        );
    }

    /// Log worker stop.
    pub fn worker_stop(&self, reason: &str) {
        self.info(
            LogEvent::WorkerStop,
            &format!("Worker stopped: {}", reason),
            LogOptions::default(),
        );
    }

    /// Log child process exit.
    pub fn process_exit(&self, exit_code: i32) {
        let level = if exit_code == 0 {
            LogLevel::Info
        } else {
            LogLevel::Warn
        };
        self.log(
            LogEvent::ProcessExit,
            &format!("Child process exited with code {}", exit_code),
            level,
            LogOptions {
                metadata: Some(serde_json::json!({ "exit_code": exit_code })),
                ..Default::default()
            },
        );
    }

    /// Log heartbeat sent.
    pub fn heartbeat_sent(&self) {
        self.debug(
            LogEvent::HeartbeatSent,
            "Heartbeat sent",
            LogOptions::default(),
        );
    }

    /// Log heartbeat response received.
    pub fn heartbeat_received(&self, rtt_ms: f64) {
        self.debug(
            LogEvent::HeartbeatReceived,
            &format!("Heartbeat received (RTT: {:.1}ms)", rtt_ms),
            LogOptions {
                duration_ms: Some(rtt_ms),
                ..Default::default()
            },
        );
    }

    /// Log missed heartbeat.
    pub fn heartbeat_missed(&self, consecutive: usize, max_allowed: usize) {
        self.warn(
            LogEvent::HeartbeatMissed,
            &format!("Heartbeat missed ({}/{})", consecutive, max_allowed),
            LogOptions {
                metadata: Some(serde_json::json!({
                    "consecutive_misses": consecutive,
                    "max_allowed": max_allowed,
                })),
                ..Default::default()
            },
        );
    }

    /// Log heartbeat timeout (too many misses).
    pub fn heartbeat_timeout(&self, misses: usize) {
        self.error(
            LogEvent::HeartbeatTimeout,
            &format!("Heartbeat timeout after {} consecutive misses", misses),
            LogOptions {
                metadata: Some(serde_json::json!({ "total_misses": misses })),
                ..Default::default()
            },
        );
    }
}

/// Options for log entries.
#[derive(Default)]
pub struct LogOptions {
    pub correlation_id: Option<String>,
    pub request_id: Option<String>,
    pub parent_request_id: Option<String>,
    pub function: Option<String>,
    pub namespace: Option<String>,
    pub duration_ms: Option<f64>,
    pub success: Option<bool>,
    pub error: Option<String>,
    pub error_type: Option<String>,
    pub metrics: Option<serde_json::Value>,
    pub metadata: Option<serde_json::Value>,
}

/// Default handler that prints JSON to stdout.
pub fn default_json_handler(entry: &LogEntry) {
    println!("{}", entry.to_json());
}

/// Default handler that prints human-readable output.
pub fn default_pretty_handler(entry: &LogEntry) {
    use std::time::{SystemTime, UNIX_EPOCH};

    let timestamp = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0);
    let time_str = format!(
        "{:02}:{:02}:{:02}",
        (timestamp / 3600) % 24,
        (timestamp / 60) % 60,
        timestamp % 60
    );
    let level = format!("{:<5}", entry.level.to_uppercase());
    let prefix = format!("[{}] [{}]", time_str, level);

    let mut parts = vec![prefix, entry.event.clone(), entry.message.clone()];

    if let Some(ref req_id) = entry.request_id {
        parts.push(format!("req={}", &req_id[..req_id.len().min(8)]));
    }
    if let Some(ref func) = entry.function {
        parts.push(format!("fn={}", func));
    }
    if let Some(duration) = entry.duration_ms {
        parts.push(format!("{:.1}ms", duration));
    }
    if let Some(ref err) = entry.error {
        parts.push(format!("error={}", err));
    }

    println!("{}", parts.join(" "));
}

fn current_timestamp() -> f64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use std::sync::{Arc, Mutex};

    #[test]
    fn test_log_level_display() {
        assert_eq!(LogLevel::Debug.to_string(), "debug");
        assert_eq!(LogLevel::Info.to_string(), "info");
        assert_eq!(LogLevel::Warn.to_string(), "warn");
        assert_eq!(LogLevel::Error.to_string(), "error");
    }

    #[test]
    fn test_log_event_display() {
        assert_eq!(LogEvent::WorkerStart.to_string(), "worker_start");
        assert_eq!(LogEvent::WorkerStop.to_string(), "worker_stop");
        assert_eq!(LogEvent::RequestStart.to_string(), "request_start");
        assert_eq!(LogEvent::CircuitOpen.to_string(), "circuit_open");
    }

    #[test]
    fn test_log_entry_to_json() {
        let entry = LogEntry {
            event: "test_event".to_string(),
            level: "info".to_string(),
            message: "Test message".to_string(),
            timestamp: 1234567890.0,
            correlation_id: Some("corr-123".to_string()),
            request_id: Some("req-456".to_string()),
            worker_id: Some("worker-1".to_string()),
            function: Some("test_func".to_string()),
            duration_ms: Some(100.5),
            success: Some(true),
            ..Default::default()
        };

        let json = entry.to_json();
        assert!(json.contains("test_event"));
        assert!(json.contains("Test message"));
        assert!(json.contains("corr-123"));
        assert!(json.contains("req-456"));
    }

    #[test]
    fn test_log_entry_default() {
        let entry = LogEntry::default();
        assert_eq!(entry.event, "");
        assert_eq!(entry.level, "");
        assert_eq!(entry.message, "");
        assert_eq!(entry.correlation_id, None);
        assert_eq!(entry.request_id, None);
    }

    #[test]
    fn test_logger_creation() {
        let logger =
            StructuredLogger::new(None, LogLevel::Info, Some("worker-1".to_string()), None);

        // Should not panic
        logger.info(LogEvent::WorkerStart, "Test", LogOptions::default());
    }

    #[test]
    fn test_logger_with_handler() {
        let logs = Arc::new(Mutex::new(Vec::new()));
        let logs_clone = Arc::clone(&logs);

        let handler = Arc::new(move |entry: &LogEntry| {
            logs_clone.lock().unwrap().push(entry.to_json());
        });

        let logger = StructuredLogger::new(
            Some(handler),
            LogLevel::Debug,
            Some("worker-1".to_string()),
            None,
        );

        logger.info(
            LogEvent::WorkerStart,
            "Worker started",
            LogOptions::default(),
        );

        let captured_logs = logs.lock().unwrap();
        assert_eq!(captured_logs.len(), 1);
        assert!(captured_logs[0].contains("worker_start"));
        assert!(captured_logs[0].contains("Worker started"));
    }

    #[test]
    fn test_log_level_filtering() {
        let logs = Arc::new(Mutex::new(Vec::new()));
        let logs_clone = Arc::clone(&logs);

        let handler = Arc::new(move |entry: &LogEntry| {
            logs_clone.lock().unwrap().push(entry.to_json());
        });

        let logger = StructuredLogger::new(
            Some(handler),
            LogLevel::Warn,
            Some("worker-1".to_string()),
            None,
        );

        // Debug and Info should be filtered out
        logger.debug(
            LogEvent::WorkerStart,
            "Debug message",
            LogOptions::default(),
        );
        logger.info(LogEvent::WorkerStart, "Info message", LogOptions::default());

        // Warn and Error should pass through
        logger.warn(LogEvent::CircuitOpen, "Warn message", LogOptions::default());
        logger.error(
            LogEvent::RequestError,
            "Error message",
            LogOptions::default(),
        );

        let captured_logs = logs.lock().unwrap();
        assert_eq!(captured_logs.len(), 2);
    }

    #[test]
    fn test_log_level_ordering() {
        assert!((LogLevel::Debug as u8) < (LogLevel::Info as u8));
        assert!((LogLevel::Info as u8) < (LogLevel::Warn as u8));
        assert!((LogLevel::Warn as u8) < (LogLevel::Error as u8));
    }

    #[test]
    fn test_set_handler() {
        let logs1 = Arc::new(Mutex::new(Vec::new()));
        let logs1_clone = Arc::clone(&logs1);

        let handler1 = Arc::new(move |_entry: &LogEntry| {
            logs1_clone.lock().unwrap().push("handler1".to_string());
        });

        let mut logger = StructuredLogger::new(
            Some(handler1),
            LogLevel::Debug,
            Some("worker-1".to_string()),
            None,
        );

        logger.info(LogEvent::WorkerStart, "Test", LogOptions::default());

        let logs2 = Arc::new(Mutex::new(Vec::new()));
        let logs2_clone = Arc::clone(&logs2);

        let handler2 = Arc::new(move |_entry: &LogEntry| {
            logs2_clone.lock().unwrap().push("handler2".to_string());
        });

        logger.set_handler(handler2);
        logger.info(LogEvent::WorkerStart, "Test", LogOptions::default());

        assert_eq!(logs1.lock().unwrap().len(), 1);
        assert_eq!(logs2.lock().unwrap().len(), 1);
    }

    #[test]
    fn test_set_context() {
        let logs = Arc::new(Mutex::new(Vec::new()));
        let logs_clone = Arc::clone(&logs);

        let handler = Arc::new(move |entry: &LogEntry| {
            logs_clone.lock().unwrap().push(entry.to_json());
        });

        let mut logger = StructuredLogger::new(Some(handler), LogLevel::Debug, None, None);

        logger.info(LogEvent::WorkerStart, "Test", LogOptions::default());

        logger.set_context(Some("worker-1".to_string()), Some("service-1".to_string()));
        logger.info(LogEvent::WorkerStart, "Test", LogOptions::default());

        let captured_logs = logs.lock().unwrap();
        assert_eq!(captured_logs.len(), 2);
        // First log should not have worker_id
        assert!(!captured_logs[0].contains("worker-1"));
        // Second log should have worker_id
        assert!(captured_logs[1].contains("worker-1"));
        assert!(captured_logs[1].contains("service-1"));
    }

    #[test]
    fn test_convenience_methods() {
        let logs = Arc::new(Mutex::new(Vec::new()));
        let logs_clone = Arc::clone(&logs);

        let handler = Arc::new(move |entry: &LogEntry| {
            logs_clone.lock().unwrap().push(entry.to_json());
        });

        let logger = StructuredLogger::new(
            Some(handler),
            LogLevel::Debug,
            Some("worker-1".to_string()),
            None,
        );

        logger.request_start(
            "req-123",
            "add",
            "default",
            Some("corr-456".to_string()),
            None,
        );
        logger.request_end("req-123", "add", 100.5, true, None, None);
        logger.circuit_open(5);
        logger.circuit_close();
        logger.worker_start("spawn");
        logger.worker_stop("shutdown");
        logger.process_exit(0);
        logger.heartbeat_sent();
        logger.heartbeat_received(15.5);
        logger.heartbeat_missed(2, 3);
        logger.heartbeat_timeout(3);

        let captured_logs = logs.lock().unwrap();
        // Note: request_end may log at different levels based on success
        assert!(captured_logs.len() >= 10);
    }

    #[test]
    fn test_log_options_default() {
        let options = LogOptions::default();
        assert!(options.correlation_id.is_none());
        assert!(options.request_id.is_none());
        assert!(options.function.is_none());
        assert!(options.namespace.is_none());
        assert!(options.duration_ms.is_none());
        assert!(options.success.is_none());
        assert!(options.error.is_none());
        assert!(options.error_type.is_none());
        assert!(options.metrics.is_none());
        assert!(options.metadata.is_none());
    }

    #[test]
    fn test_log_options_with_fields() {
        let options = LogOptions {
            correlation_id: Some("corr-123".to_string()),
            request_id: Some("req-456".to_string()),
            parent_request_id: Some("parent-789".to_string()),
            function: Some("test_func".to_string()),
            namespace: Some("test_ns".to_string()),
            duration_ms: Some(100.5),
            success: Some(true),
            error: Some("Test error".to_string()),
            error_type: Some("TestError".to_string()),
            metrics: Some(json!({"count": 1})),
            metadata: Some(json!({"key": "value"})),
        };

        assert_eq!(options.correlation_id, Some("corr-123".to_string()));
        assert_eq!(options.request_id, Some("req-456".to_string()));
        assert_eq!(options.function, Some("test_func".to_string()));
    }

    #[test]
    fn test_default_json_handler() {
        let entry = LogEntry {
            event: "test_event".to_string(),
            level: "info".to_string(),
            message: "Test message".to_string(),
            timestamp: 1234567890.0,
            ..Default::default()
        };

        // Should not panic
        default_json_handler(&entry);
    }

    #[test]
    fn test_default_pretty_handler() {
        let entry = LogEntry {
            event: "test_event".to_string(),
            level: "info".to_string(),
            message: "Test message".to_string(),
            timestamp: 1234567890.0,
            request_id: Some("req-123".to_string()),
            function: Some("test_func".to_string()),
            duration_ms: Some(100.5),
            ..Default::default()
        };

        // Should not panic
        default_pretty_handler(&entry);
    }

    #[test]
    fn test_log_entry_serialization() {
        let entry = LogEntry {
            event: "test_event".to_string(),
            level: "info".to_string(),
            message: "Test message".to_string(),
            timestamp: 1234567890.0,
            correlation_id: Some("corr-123".to_string()),
            request_id: Some("req-456".to_string()),
            parent_request_id: Some("parent-789".to_string()),
            worker_id: Some("worker-1".to_string()),
            service_id: Some("service-1".to_string()),
            function: Some("test_func".to_string()),
            namespace: Some("test_ns".to_string()),
            duration_ms: Some(100.5),
            success: Some(true),
            error: Some("Test error".to_string()),
            error_type: Some("TestError".to_string()),
            metrics: Some(json!({"count": 1})),
            metadata: Some(json!({"key": "value"})),
        };

        let json = entry.to_json();
        let parsed: serde_json::Value = serde_json::from_str(&json).unwrap();

        assert_eq!(parsed["event"], "test_event");
        assert_eq!(parsed["level"], "info");
        assert_eq!(parsed["message"], "Test message");
        assert_eq!(parsed["correlation_id"], "corr-123");
        assert_eq!(parsed["request_id"], "req-456");
        assert_eq!(parsed["worker_id"], "worker-1");
        assert_eq!(parsed["service_id"], "service-1");
        assert_eq!(parsed["function"], "test_func");
        assert_eq!(parsed["namespace"], "test_ns");
        assert_eq!(parsed["duration_ms"], 100.5);
        assert_eq!(parsed["success"], true);
        assert_eq!(parsed["error"], "Test error");
        assert_eq!(parsed["error_type"], "TestError");
    }

    #[test]
    fn test_handler_exception_handling() {
        // Handler that panics should not crash the logger
        let handler = Arc::new(|_entry: &LogEntry| {
            panic!("Handler panic!");
        });

        let logger = StructuredLogger::new(
            Some(handler),
            LogLevel::Debug,
            Some("worker-1".to_string()),
            None,
        );

        // This should not panic (the panic is caught by the handler's closure)
        // Note: In Rust, panics in closures will propagate unless caught
        // This test verifies the structure is correct
        assert_eq!(logger.worker_id, Some("worker-1".to_string()));
    }
}
